src/world.cpp

```diff
@@ -78,33 +122,14 @@ void world_act_npcs(void *world_void)
 	{
 		UTIL_FOREACH(map->npcs, npc)
 		{
-			const auto& data = npc->Data();
-
 			if (npc->alive)
 			{
 				if (npc->last_act + npc->act_speed < current_time)
 				{
 					npc->Act();
 				}
 
-				int chance = util::rand(0, 99);
-				if (npc->last_talk + data.talk_speed < current_time &&
-					chance < data.talk_chance &&
-					data.speech.size() > 0)
-				{
-					int ndx = util::rand(0, data.speech.size() - 1);
-					std::string speech = data.speech[ndx];
-					npc->map->Msg(npc, speech);
-					npc->last_talk += data.talk_speed;
-
-					UTIL_FOREACH(npc->map->npcs, npciter)
-					{
-						if (npciter->spawn_x == npc->spawn_x && npciter->spawn_y == npc->spawn_y)
-						{
-							npciter->last_talk = npc->last_talk;
-						}
-					}
-				}
+				world_npc_speech(npc);
```

Comment:
Importantly, the call to `world_npc_speech` isn't occurring within the `last_act` check, so NPCs can speak but not act in the same tick. If we put NPC speech directly into `NPC::Act` then NPCs could only speak on the same tick as an act, which could introduce noticeable delay for NPCs with slower speeds and cause our speech frequency to be wrong. That seems undesirable to me.
We're currently only piggy-backing onto the NPC act timer because it's an existing hot loop over all all NPCs that fires every 50ms.
We could probably just add a separate 1s timer that loops over all NPCs for speech checks, and then drag the majority of `world_npc_speech` into an `NPC::Speak`` function. How does that sound?

Reply:
Yeah I like that idea, I'd go with that.

Reply:
If you do go that route, don't forget the `npc->alive` check, since that is done out of the scope of `world_npc_speech`